kvserver: use combined store/repl critical section in applySnapshot

tbg opened the pull request:
This PR addresses #58378, which highlighted insufficient locking in `(*Replica).applySnapshot`. Prior to this PR, `applySnapshot` carried out operations in roughly this order:

1. ingest SSTs into the LSM
2. update the replica's descriptor and the store's map of span to replica
3. update the replica state

There was no critical section wrapping both of 2) and 3). As a result, it
was possible (in theory, not observed) to obtain a replica from the store
under descriptor `X` whose state had not entirely been updated to reflect `X`.

It's hard to say where exactly you would see issues with this, but there was
one instance that caused a crash: when a replica was created in response to
an incoming snapshot, its state was completely empty; such a replica is not
initialized; they are never handed out by the store; they don't have anything
in their state.

Due to the race described above, the store would sometimes hand out these
replicas when their descriptor had just been initialized, but its state
not populated yet, which caused at least two crashes related to calls to
`(*Replica).Version`:

1. from `PurgeOutdatedReplicas`, and
2. during evaluation of a lease request, triggering [this assertion]

[this assertion]: ef645191867f2ba4d80083f1dae0b9ba1b805277

Both previously had workarounds in place to avoid the crash, which are now removed.

Release note: None

tbg committed:
kvserver: hold r.mu across leasePostApplyLocked

This is step one towards a solution to #58378. We want to update the
in-memory range state in the same critical section as the store
addressing update, and this includes calling leasePostApplyLocked.

As an interim step, this commit moves the call into a large `r.mu`
critical section that we will later combine with the `s.mu` critical
section during `applySnapshot`.

Release note: None

tbg committed:
kvserver: use combined store/repl critical section in applySnapshot

Fixes #58378.

Release note: None

tbg committed:
kvserver: remove obsolete stop-gaps

These were introduced to temporarily mitigate #58378, which the
preceding commit fixed. On my machine,

```
make stress PKG=./pkg/server/ TESTS=TestMigrationPurgeOutdatedReplicas STRESSFLAGS='-p 8'
```

failed without the preceding two locking changes, in 51s, 3m30s, and
6m44s, respectively. With the changes, it did not fail within 15m:

> 41944 runs so far, 0 failures, over 15m0s

Release note: None

tbg committed:
kvserver: re-arm the truncated state assertion

This reverts ef645191867f2ba4d80083f1dae0b9ba1b805277, which should
now be fine since the preceding commits fixed #58378.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59423)
<!-- Reviewable:end -->

tbg commented:
There is still some locking nightmare solving necessary here. Don't look yet.

tbg commented:
The thing I missed is that as of right now, inserting into the `replicasByKey` btree acquires the mutex (to call `.Less()`). But I just changed it such that the mutex is held across the insert operation, so there's a deadlock.

This seems silly though, all the btree needs is the replica start key, which is immutable (for initialized replicas, which is all that matters here). So - I will make `*Replica`'s impl of `btree.Item` non-locking, but there is a bunch of fallout. It's the good kind of fallout - should simplify matters - but it's fallout. I'll spend some extra time here as this is some of the most inscrutable code I've had to touch in a while.

tbg committed:
kvserver: use immutable startKey in replicasByKey

This allows us to have `*Replica` implement `btree.Item` with a
lockless `Less` method.

Release note: None

tbg commented:
RFAL!

irfansharif reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> looks like this was pretty hairy to untangle. 

> Reviewed 6 of 6 files at r1, 6 of 6 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-:-MSTtLnJ4xw6og8b2lwZ:b5x0stn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSTqpV98pMS0uovCfKG:-MSTqpVAErmI7gjg533G:b7exd3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa0d422ce588e3f4e73b35696bede89e5aed9b68/pkg/kv/kvserver/replica_proposal.go#L401)):*
> ```Go
> 
> 	// Inform the concurrency manager that the lease holder has been updated.
> 	r.concMgr.OnRangeLeaseUpdated(newLease.Sequence, iAmTheLeaseHolder)
> ```

Is there a reason we re-ordered this chunk? (I assume the original ordering was not required for safety, but can't tell off-hand.)
___
*[pkg/kv/kvserver/store_replica_btree_test.go, line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSTsNaa4At9AuMrmulZ:-MSTsNaa4At9AuMrmul_:bdhy4jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ebd3dd1e70e51b1aa902ab73fb75e96682cbef/pkg/kv/kvserver/store_replica_btree_test.go#L100)):*
> ```Go
> 		r := &Replica{}
> 		r.mu.state.Desc = desc
> 		r.startKey = desc.StartKey // this is what's actually used in btree
> ```

[nit] "the btree"


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 6 files at r1, 6 of 6 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59423)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSTqpV98pMS0uovCfKG:-MSV7vZUAb88rjjbp-sm:b-r3x81o) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa0d422ce588e3f4e73b35696bede89e5aed9b68/pkg/kv/kvserver/replica_proposal.go#L401)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is there a reason we re-ordered this chunk? (I assume the original ordering was not required for safety, but can't tell off-hand.)
</blockquote></details>

Yeah, I'm also curious about this. I think the original ordering was required for safety, as we didn't want to have a range with a lease whose concurrent manager was disabled. With the new locking, it may not matter.
___
*[pkg/kv/kvserver/replica_proposal.go, line 784 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSV9SuJ8BKnZR36QBIY:-MSV9SuK7G-R534wHhBJ:bdeavht) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2562d4df735f5c6b52374867d5fd41ae5cf2c10/pkg/kv/kvserver/replica_proposal.go#L784)):*
> ```Go
> 			migrationVersion := clusterversion.ByKey(clusterversion.TruncatedAndRangeAppliedStateMigration)
> 			if migrationVersion.Less(activeVersion) {
> 				log.Fatal(ctx, "not using applied state key in v21.1")
> ```

:fire: 
___
*[pkg/kv/kvserver/store_create_replica.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSV8ozLE4-3ogq2tmZv:-MSV8ozLE4-3ogq2tmZw:brnapc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ebd3dd1e70e51b1aa902ab73fb75e96682cbef/pkg/kv/kvserver/store_create_replica.go#L320)):*
> ```Go
> 
> 	rangeID := repl.RangeID
> 	repl.startKey = repl.Desc().StartKey
> ```

I know `repl.startKey`'s value will never change once set, but does that mean that it can be accessed outside the replica lock while we overwrite the slice with an identical one under the replica lock? Will the race detector complain?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
It looks like there's a new test flake in `TestReplicaDrainLease` on this PR which I need to investigate. Unfortunately, it's not an easy flake to hit (though CI did hit it). Basically RFAL though

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-:-MSXIIMA5wEUm1Rqr3MY:beox4au)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSTqpV98pMS0uovCfKG:-MSWvxdn4MkYpcj1sw_x:b-58iyju) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa0d422ce588e3f4e73b35696bede89e5aed9b68/pkg/kv/kvserver/replica_proposal.go#L401)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I'm also curious about this. I think the original ordering was required for safety, as we didn't want to have a range with a lease whose concurrent manager was disabled. With the new locking, it may not matter.
</blockquote></details>

Interesting, I reordered this because I got worried about the opposite, namely that the concurrency manager would ping some commands and they would see the old lease instead of the new. I reordered this back, with a comment.
___
*[pkg/kv/kvserver/store_create_replica.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSV8ozLE4-3ogq2tmZv:-MSWxcfGAFHfh7c029-5:b-sj8jbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ebd3dd1e70e51b1aa902ab73fb75e96682cbef/pkg/kv/kvserver/store_create_replica.go#L320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I know `repl.startKey`'s value will never change once set, but does that mean that it can be accessed outside the replica lock while we overwrite the slice with an identical one under the replica lock? Will the race detector complain?
</blockquote></details>

We never overwrite `startKey`, we set it only once (both writes only happen when the repl is not initialized yet). Effectively, we are using `r.mu` for locking here. As long as `r` is uninitialized, you're not allowed to access `startKey`, and we don't. To know that the replica is initialized, you have to acquire `r.mu` and observe the initialization. But if you do that, and you see that `r` is initialized, `r.startKey` has already been written (as it is written under the same `r.mu` that made the replica initialized). Is this too subtle? I can slap a one-off mutex around this.


<!-- Sent from Reviewable.io -->

tbg commented:
Jeez, that's going to be fun:

```
=== RUN   TestReplicaDrainLease
    test_log_scope.go:73: test logs captured to: /tmp/logTestReplicaDrainLease596883025
    test_log_scope.go:74: use -show-logs to present logs inline
    replica_test.go:1524: 
        	Error Trace:	replica_test.go:1524
        	Error:      	Received unexpected error:
        	           	[NotLeaseHolderError] lease state couldn't be determined: lease:<start:<> replica:<node_id:1 store_id:1 replica_id:1 > proposed_ts:<wall_time:1612269138271636714 > epoch:1 sequence:1 > now:<wall_time:1612269140497806811 > request_time:<wall_time:1612269140497806811 > liveness:<node_id:1 expiration:<wall_time:0 logical:0 > > ; r36: replica (n2,s2):2 not lease holder; lease holder unknown
        	Test:       	TestReplicaDrainLease
    panic.go:617: -- test log scope end --
test logs left over in: /tmp/logTestReplicaDrainLease596883025
--- FAIL: TestReplicaDrainLease (2.29s)
FAIL


ERROR: exit status 1

26984 runs completed, 1 failures, over 1h7m26s
```

tbg commented:
Ok, I think I have the test failure fixed.

tbg commented:
Friendly ping. Another look here can't hurt.

irfansharif commented on pkg/kv/kvserver/replica.go:
```diff
@@ -1106,6 +1104,11 @@ func (r *Replica) assertStateLocked(ctx context.Context, reader storage.Reader)
 	}
+	if r.isInitializedRLocked() {
+		if !r.startKey.Equal(r.mu.state.Desc.StartKey) {
+			log.Fatalf(ctx, "denormalized start key %s diverged from %s", r.startKey, r.mu.state.Desc.StartKey)
```
What's denormalized about it?

irfansharif reviewed (approved):
Looks like we have a few logictests to update.

> As long as `r` is uninitialized, you're not allowed to access `startKey`, and we don't.

I'm guessing the race detector would complain if we ever tried it?

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 6 files at r7, 5 of 5 files at r8, 1 of 1 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-:-MScuqjvAmnc47AnyGDG:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/kv/kvserver/store_create_replica.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSV8ozLE4-3ogq2tmZv:-MSct1MO9G1GVq4ETiwm:b-ay5f3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ebd3dd1e70e51b1aa902ab73fb75e96682cbef/pkg/kv/kvserver/store_create_replica.go#L320)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We never overwrite `startKey`, we set it only once (both writes only happen when the repl is not initialized yet). Effectively, we are using `r.mu` for locking here. As long as `r` is uninitialized, you're not allowed to access `startKey`, and we don't. To know that the replica is initialized, you have to acquire `r.mu` and observe the initialization. But if you do that, and you see that `r` is initialized, `r.startKey` has already been written (as it is written under the same `r.mu` that made the replica initialized). Is this too subtle? I can slap a one-off mutex around this.
</blockquote></details>

I see. This seems fine then. But then consider two small improvements to make this a little more clear:
1. add a bit of what you said here to the comment on the field. Specifically, the part about how observing the initialization requires `r.mu`.
2. don't set `startKey` directly. Instead, create a method (`setStartKeyLocked` or something) that asserts that the startKey has not already been set.


<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: clarify semantics of Replica.startKey

Release note: None

tbg reviewed (commented):
> I'm guessing the race detector would complain if we ever tried it?

Probably true.

> Dismissed @irfansharif from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-:-MSdRTTb5PGS2wWz_7pC:b-swy0xp)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica.go, line 1109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MRyI7vCo18OdYSlSlIo-r10-1109:-MSdR_iL0c4V3QJRdYQf:b607t4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b9a8db37bbffed4e55159286ca58b10633da010/pkg/kv/kvserver/replica.go#L1109)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

What's denormalized about it?
</blockquote></details>

The authoritative one is `r.mu.state.Desc.StartKey`, "denormalization" is used here loosely to mean "a redundant copy"
___
*[pkg/kv/kvserver/store_create_replica.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59423#-MSV8ozLE4-3ogq2tmZv:-MSdTRN-2up09Rubvlbr:b-a93i7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ebd3dd1e70e51b1aa902ab73fb75e96682cbef/pkg/kv/kvserver/store_create_replica.go#L320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I see. This seems fine then. But then consider two small improvements to make this a little more clear:
1. add a bit of what you said here to the comment on the field. Specifically, the part about how observing the initialization requires `r.mu`.
2. don't set `startKey` directly. Instead, create a method (`setStartKeyLocked` or something) that asserts that the startKey has not already been set.
</blockquote></details>

Good ideas, done.


<!-- Sent from Reviewable.io -->

tbg commented:
The test flake is https://github.com/cockroachdb/cockroach/pull/59802.

bors r=nvanbenschoten,irfansharif
